### PR TITLE
Improve TS declarations with no breaking change

### LINKED ISSUE #347: `JSJoda.Instant is not assignable to parameter of type 'import (.../core/dist/js-joda).Instant`
the new published extra package seems to be missing Instant definitions from core, so its unusable.

### PR DESCRIPTION
**Pull request by InExtremaRes**
Improve TypeScript declarations without requiring a major release.

It does **not** supersede #353. It takes some of the changes of that PR but only those that are safe and don't make a breaking change. My intention is, if this PR is accepted and merged, rebase #353 so it still can be applied for a major release with presumably other breaking changes.

This PR adds missing methods to `Year`, remove some `any` types, fix some `TemporalUnit` previously typed as `IsoField` and fix the types of `@js-joda/extra` without changing to ES module.

It removes the `LocalDateTime.of` that takes zero arguments, now requiring at least three. This is not a breaking change since the call with zero arguments fails at runtime anyway.

Fixes #347

### CODE REVIEW

**Comment by mg-vmoaero** [comment]
If this gets approved you'll have made my current project at work significantly easier so I'd like to thank you for that 😁.

**Comment by InExtremaRes** [comment]
@mg-vmoaero Thank you.

I'm testing this in my project and it works. You can trying this changes just copying both .d.ts files from core and extra and replacing those on your `node_modules` folder. Of course is a temporary solution (any `npm install` could replace the files again) but it will be very valuable feedback and we could know if this really solves your problem. Let me know if you can test this.

**Comment by mg-vmoaero** [comment]
> @mg-vmoaero Thank you.
> 
> I'm testing this in my project and it works. You can trying this changes just copying both .d.ts files from core and extra and replacing those on your `node_modules` folder. Of course is a temporary solution (any `npm install` could replace the files again) but it will be very valuable feedback and we could know if this really solves your problem. Let me know if you can test this.

unfortunately I just don't have the time to test this atm, crunch time.

File: packages/core/dist/js-joda.d.ts [hunk 1]
```diff
@@ -959,7 +971,7 @@ declare namespace JSJoda {
 
         isBefore(other: LocalDateTime): boolean
 
-        isEqual(other: any): boolean
+        isEqual(other: LocalDateTime): boolean
```
[the remark this answers is not in the document]
Comment (InExtremaRes) [reply packages/core/dist/js-joda.d.ts]: `isEqual`, unlike `equals`, expect the argument to be of the same instance:

```js
LocalDateTime.now().isEqual(LocalDate.now());
```

fails with:

```plain
IllegalArgumentException: other must be an instance of LocalDateTime, but is LocalDate
```

so not really accepts _any_, and is now consistent with `isBefore` and `isAfter`.

File: packages/core/dist/js-joda.d.ts [hunk 2]
```diff
@@ -763,15 +770,20 @@ declare namespace JSJoda {
...
-    class IsoFields {
+    class IsoFields extends TemporalField {
         static DAY_OF_QUARTER: IsoFields
         static QUARTER_OF_YEAR: IsoFields
         static WEEK_OF_WEEK_BASED_YEAR: IsoFields
         static WEEK_BASED_YEAR: IsoFields
-        static WEEK_BASED_YEARS: IsoFields
-        static QUARTER_YEARS: IsoFields
+        static WEEK_BASED_YEARS: TemporalUnit
+        static QUARTER_YEARS: TemporalUnit
```
[the remark this answers is not in the document]
Comment (InExtremaRes) [reply packages/core/dist/js-joda.d.ts]: In theory and just from a type perspective, yes it is. But they're not really `IsoFields` and trying to using them as such fails at runtime. For instance:

```js
const ld = LocalDate.now();

ld.get(IsoFields.QUARTER_YEARS);
// TypeError: field.getFrom is not a function

ld.plus(1, IsoFields.QUARTER_YEARS);
// Works at runtime, fails to compile
```

So they cannot be used as is. I think this counts as a bug fix.

File: packages/core/dist/js-joda.d.ts [hunk 3]
```diff
@@ -763,15 +770,20 @@ declare namespace JSJoda {
...
 
         private constructor()
+
+        isDateBased(): boolean
+        isTimeBased(): boolean
+        name(): string
+        getDisplayName(): string
     }
```
[the remark this answers is not in the document]
Comment (InExtremaRes) [reply packages/core/dist/js-joda.d.ts]: Your are right, the change should've been to change the static properties to `TemporalField`. I'll fix it.

File: packages/core/dist/js-joda.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pithu) [reply packages/core/dist/js-joda.d.ts]: You are right, i oversee the `is`

**Comment by InExtremaRes** [comment]
@pithu `IsoFields` does not longer extends from `TemporalField`.

About the "breaking" changes, you're right that those in #353 are not that dramatic. However, there are more changes I would like to see, not only to types but also to names that not that consistent; I will present some more ideas in #355 to discussion and those will deserve a major release if you accept them. Maybe it is not a problem to make a major release now and then another if everything goes well?

Anyhow, I think this PR could still be valuable for those than cannot afford a major update right now and includes some bug fixes for TS users.

**Pull request merged**